File: Either.java
public abstract class Either<A, B> {
    public boolean isRight;

    public abstract <C> Either<A, C> fmap(Function<B, C> f);
}

class Right<A, B> extends Either<A, B> {
    public B val;
    public Right(B b){
        val = b;
        isRight = true;
    }

    public <C> Either<A, C> fmap(Function<B, C> f){
        return new Right<A, C>(f.apply(val));
    }

    public <C> Either<A, C> bind(Function<B, Either<A, C>> f){
        return f.apply(val);
    }
}

class Left<A, B> extends Either<A, B> {
    public final A val;
    public Left(A a){
        val = a;
        isRight = false;
    }

    public <C> Either<A, C> fmap(Function<B, C> f){
        return new Left<A, C>(val);
    }

    public <C> Either<A, C> bind(Function<B, Either<A, C>> f){
        return new Left<A, C>(val);
    }
}

File: csharp/Monads/Either.cs
using System;

namespace Monads
{
	namespace Either
	{
		public abstract class Either<L, R>
		{
			public abstract Either<L, X> FMap<X>(Func<R, X> f);
			public abstract Either<L, X> Bind<X>(Func<R, Either<L, X>> f);
		}

		public class Right<L, R> : Either<L, R>
		{
			public readonly R r;
			public Right(R r)
			{
				this.r = r;
			}
			public override Either<L, X> FMap<X>(Func<R, X> f){
				return new Right<L, X>(f.Invoke (this.r));
			}
			public override Either<L, X> Bind<X>(Func<R, Either<L, X>> f)
			{
				return f.Invoke (this.r);
			}
		}

		public class Left<L, R> : Either<L, R>
		{
			public readonly L l;
			public Left(L l)
			{
				this.l = l;
			}
			public override Either<L, X> FMap<X>(Func<R, X> f)
			{
				return new Left<L, X> (this.l);
			}
			public override Either<L, X> Bind<X>(Func<R, Either<L, X>> f)
			{
				return new Left<L, X> (this.l);
			}
		}
	}
}

